File: C++/UniqueBinarySearchTreesII.cpp
#include<iostream>
#include<vector>
#include<string>
#include<string.h>
#include<math.h>
#include<algorithm>
using namespace std;
struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	vector<TreeNode *> gen(int s, int e)
	{
		vector<TreeNode *> ret;
		if(s > e)
		{
			ret.push_back(NULL);
			return ret;
		}
		for(int i = s; i <= e; i++)
		{
			vector<TreeNode*> lefts = gen(s,i-1);
			vector<TreeNode*> rights = gen(i+1, e);
			for(int k = 0; k < lefts.size(); k++)
			{
				for(int j = 0; j < rights.size(); j++)
				{
					TreeNode* root = new TreeNode(i);
					root->left = lefts[k];
					root->right = rights[j];
					ret.push_back(root);
				}
			}
		}
		return ret;
	}
	vector<TreeNode *> generateTrees(int n) {
		// Start typing your C/C++ solution below
		// DO NOT write int main() function
		return gen(1, n);
	}
};

File: JAVA/UniqueBinarySearchTreesII.java
import java.util.ArrayList;

class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;

	TreeNode(int x) {
		val = x;
		left = null;
		right = null;
	}
}

class Solution {
	public ArrayList<TreeNode> generateTrees(int n) {
		// Start typing your Java solution below
		// DO NOT write main() function
		return generate(1, n);
	}

	private ArrayList<TreeNode> generate(int start, int end) {
		ArrayList<TreeNode> ret = new ArrayList<TreeNode>();
		if(start > end)
		{
			ret.add(null);
			return ret;
		}
		for(int i = start; i <= end; i++)
		{
			ArrayList<TreeNode> left = generate(start, i-1);
			ArrayList<TreeNode> right = generate(i+1, end);
			for(int j = 0; j < left.size(); j++)
			{
				for(int k = 0; k < right.size(); k++)
				{
					TreeNode root = new TreeNode(i);
					root.left = left.get(j);
					root.right = right.get(k);
					ret.add(root);
				}
			}
		}
		return ret;
	}
}

public class UniqueBinarySearchTreesII {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		// TODO Auto-generated method stub

	}

}
